pkg/apis/configs/v1alpha1/configmappropagation_types.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/kmeta"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ConfigMapPropagation is used to propagate configMaps from original namespace to current namespace
+type ConfigMapPropagation struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the ConfigMapPropagation
+	Spec ConfigMapPropagationSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the EventType.
+	// This data may be out of date.
+	// +optional
+	Status ConfigMapPropagationStatus `json:"status,omitempty"`
+}
+
+var (
+	// Check that ConfigMapPropagation can be validated, can be defaulted, and has immutable fields.
+	_ apis.Validatable = (*ConfigMapPropagation)(nil)
+	_ apis.Defaultable = (*ConfigMapPropagation)(nil)
+
+	// Check that ConfigMapPropagation can return its spec untyped.
+	_ apis.HasSpec = (*ConfigMapPropagation)(nil)
+
+	_ runtime.Object = (*ConfigMapPropagation)(nil)
+
+	// Check that we can create OwnerReferences to a ConfigMapPropagation.
+	_ kmeta.OwnerRefable = (*ConfigMapPropagation)(nil)
+)
+
+type ConfigMapPropagationSpec struct {
+	// OriginalNamespace is the namespace where the original configMaps are in
+	OriginalNamespace string `json:"originalNamespace,omitempty"`
+	// Selector only selects original configMaps with corresponding labels
+	// +optional
+	Selector *map[string]string `json:"selector,omitempty"`
```

Comment:
what about using LabelSelector no avoid surprises: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#labelselector-v1-meta

Reply:
+1
Also maps _are_ pointers already.

Reply:
:+1: to using `*metav1.LabelSelector` as used in [DeploymentSpec](https://github.com/kubernetes/api/blob/9bd6c7846cc9c14a6440b78ad9f66fc36769473e/apps/v1/types.go#L279) and others in the apps group.

Reply:
@antoineco In API types, fields like this are pointers to create a distinction between a missing value and an empty value. For LabelSelector specifically this distinction has semantic meaning: ["An empty label selector matches all objects. A null label selector matches no objects."](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#labelselector-v1-meta)

Reply:
@grantr right, but _missing_ values are usually expected to be set by a defaulter (e.g. Replicas, TerminationGracePeriod, ...). Is that the case here?

Reply:
I'm only interpreting the docs; I have no inside knowledge of Kubernetes API design. :)

In some cases like the ones you mention, the missing value is a signal to the defaulting webhook that a default should be set, and the controller would not be expected to encounter an object with a missing value.

But in the case of the LabelSelector, a defaulter would not work because replacing a missing value with an empty value would convert the selector's meaning from "no objects" to "all objects" (as documented). The controller is expected to interpret a missing label selector differently from a present-but-empty label selector.

Does this make sense? I could be reading the docs wrong...

Reply:
It does yes, my point was more about the rejection than the defaulting. Maps already have this distinction between nil and empty built into the Go language, that's why I was saying a pointer to a map was error prone.
In the case of a struct, I agree we should absolutely use a pointer.

Reply:
Ah I'm sorry, I understand now! Good point about maps already allowing null and empty. ObjectMeta is an example with labels and annotations being (optional) maps:
https://github.com/kubernetes/apimachinery/blob/fb4ce0f5343808fd521a80b5dc63e535c357039f/pkg/apis/meta/v1/types.go#L231